### PR TITLE
if reconfigure hook exists, don't send SIGTERM

### LINKED ISSUE #987: Be able to specify that a service wants a SIGHUP for reconfigure
Right now the supervisor always sends a `SIGTERM` during a configuration update.  As @fnichol  said:

```
 i've love a way in a `plan.sh` to tell the supervisor that my service just needs a SIGHUP to reconfigure
```

See below for an example with prometheus where I've written a reconfigure hook to send a HUP to the child.  The process gets restarted anyway.

```
hab-sup(MN): Starting jamesc/prometheus
hab-sup(GS): Supervisor 172.17.0.2: faf98ca6-967a-4060-9b39-c57b1ad959f1
hab-sup(GS): Census prometheus.default: 7ba2c7af-bf3e-4eaa-8897-5f7d2416abec
hab-sup(GS): Starting inbound gossip listener
hab-sup(GS): Starting outbound gossip distributor
hab-sup(GS): Starting gossip failure detector
hab-sup(CN): Starting census health adjuster
hab-sup(SC): Updated prometheus.yml
hab-sup(TP): Restarting because the service config was updated via the census
prometheus(SV): Starting
prometheus(O): time="2016-06-24T16:26:51Z" level=info msg="Starting prometheus (version=0.20.0, branch=master, revision=aeab25c)
" source="main.go:73"
prometheus(O): time="2016-06-24T16:26:51Z" level=info msg="Build context (go=go1.6.2, user=root@77050118f904, date=20160616-08:3
8:14)" source="main.go:74"
prometheus(O): time="2016-06-24T16:26:51Z" level=info msg="Loading configuration file /hab/svc/prometheus/config/prometheus.yml" source="main.go:206"
prometheus(O): time="2016-06-24T16:26:51Z" level=info msg="Loading series map and head chunks..." source="storage.go:341"
prometheus(O): time="2016-06-24T16:26:51Z" level=info msg="465 series loaded." source="storage.go:346"
prometheus(O): time="2016-06-24T16:26:51Z" level=warning msg="No AlertManagers configured, not dispatching any alerts" source="notifier.go:174"
prometheus(O): time="2016-06-24T16:26:51Z" level=info msg="Starting target manager..." source="targetmanager.go:74"
prometheus(O): time="2016-06-24T16:26:51Z" level=info msg="Listening on :9090" source="web.go:241"
Writing new file from gossip: /hab/svc/prometheus/gossip.toml
hab-sup(SC): Updated prometheus.yml
prometheus(SV): Stopping
prometheus(O): time="2016-06-24T16:27:02Z" level=warning msg="Received SIGTERM, exiting gracefully..." source="main.go:188"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="See you next time!" source="main.go:195"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Stopping target manager..." source="targetmanager.go:88"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Stopping rule manager..." source="manager.go:412"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Rule manager stopped." source="manager.go:418"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Stopping notification handler..." source="notifier.go:309"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Stopping local storage..." source="storage.go:364"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Stopping maintenance loop..." source="storage.go:366"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Maintenance loop stopped." source="storage.go:1034"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Stopping series quarantining..." source="storage.go:370"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Series quarantining stopped." source="storage.go:1445"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Stopping chunk eviction..." source="storage.go:374"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Chunk eviction stopped." source="storage.go:854"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Checkpointing in-memory metrics and chunks..." source="persistence.go
:539"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Done checkpointing in-memory metrics and chunks in 4.142657ms." sourc
e="persistence.go:563"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Checkpointing fingerprint mappings..." source="persistence.go:1355"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Done checkpointing fingerprint mappings in 1.756613ms." source="persi
stence.go:1378"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Local storage stopped." source="storage.go:389"
hab-sup(SV): prometheus - process 6605 died with exit code 0
hab-sup(SV): prometheus - Service exited
prometheus(SV): Starting
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Starting prometheus (version=0.20.0, branch=master, revision=aeab25c)
" source="main.go:73"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Build context (go=go1.6.2, user=root@77050118f904, date=20160616-08:3
8:14)" source="main.go:74"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Loading configuration file /hab/svc/prometheus/config/prometheus.yml"
 source="main.go:206"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Loading series map and head chunks..." source="storage.go:341"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="465 series loaded." source="storage.go:346"
prometheus(O): time="2016-06-24T16:27:02Z" level=warning msg="No AlertManagers configured, not dispatching any alerts" source="n
otifier.go:174"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Listening on :9090" source="web.go:241"
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Starting target manager..." source="targetmanager.go:74"
hab-sup(TP): Forwarding SIGHUP on to the supervised process
prometheus(O): time="2016-06-24T16:27:02Z" level=info msg="Loading configuration file /hab/svc/prometheus/config/prometheus.yml"
 source="main.go:206"
```

Note the last line where the newly created process reloads the config once it gets the HUP.  It looks like the service is restarted, then the `reconfigure` hook is sent.

### PR DESCRIPTION
**Pull request by srenatus**
There already is a comment saying that this could do with some refactoring; for now only trying to fix the bug that @jamesc reported.

Fixes #987.

Signed-off-by: Stephan Renatus <srenatus@chef.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
@srenatus, thanks for your PR! By analyzing the annotation information on this pull request, we identified @adamhjk, @reset, @metadave, @fnichol and @juliandunn to be potential reviewers

File: components/sup/src/topology/mod.rs [hunk 1]
```diff
@@ -331,10 +331,12 @@ fn run_internal<'a>(sm: &mut StateMachine<State, Worker<'a>, SupError>,
                         // Write the configuration, and restart if needed
                         if try!(service_config.write(&package)) {
                             try!(package.copy_run(&service_config));
-                            try!(package.reconfigure(&service_config));
                             outputln!("Restarting because the service config was updated via the \
                                        census");
-                            restart_process = true;
+                            let has_reconfigure = try!(package.reconfigure(&service_config));
+                            if !has_reconfigure {
```
Comment (metadave) [inline components/sup/src/topology/mod.rs]: I like how you've pulled out `has_reconfigure` instead of lumping it all together on one line.

**Comment by metadave** [comment]
I'd like to write a spec for this before merging, I'll get there "soon".

**Comment by metadave** [comment]
closed by #1336

**Pull request closed without merging**